### PR TITLE
Import GitLab projects

### PR DESCRIPTION
**Pull request by AlexTugarev**
This PR enables importing of GitLab repositories as projects into Gitpod.

Your GitLab account needs to have at the access level of a maintainer in order to install webhooks for prebuilds. That's also a filter criteria for repositories to be choose from.

Screen Shot 2021-08-09 at 17 06 51

Closes https://github.com/gitpod-io/gitpod/issues/5278

### CODE REVIEW

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **alextugarev** after the PR has been reviewed.

*No associated issue*. Update pull-request body to add a reference to an issue, or get approval with `/approve no-issue`

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (AlexTugarev) [inline components/server/ee/src/gitlab/gitlab-app-support.ts:14]: The separation in EE and non-EE  and the scope of the `GitLabService` binding makes it impossible to use that for the additions, unfortunately.

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (gtsiolis) [inline components/server/ee/src/gitlab/gitlab-app-support.ts:19]: **issue**: This is probably minor for this iteration, as this is an edge case, but it takes 5-6 seconds for me to load the next step. FWIW, I'm a member of ~30 groups. In the future a loading indicator could help.

File: components/server/ee/src/container-module.ts
Comment (gtsiolis) [inline components/server/ee/src/container-module.ts:73]: **issue**: Adding this here although probably not related to this piece of code. Is it known that _Branches_ and _Prebuilds_ are not listed after adding a proejct?

**Review by gtsiolis — COMMENTED** [review]
Thank for pushing this forward @AlexTugarev! 🍊 

Not sure if this was ready to review but left some early comments below. 🍋

**Comment by jankeromnes** [comment]
Removing the automatic review request because this is still Draft. 😊

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (AlexTugarev) [reply components/server/ee/src/gitlab/gitlab-app-support.ts:19]: Thank for the hint. Just addmin `perPage: 100` to override the default value of `20`. That should be significant improvement in your case.

File: components/dashboard/src/projects/Project.tsx
Comment (AlexTugarev) [inline components/dashboard/src/projects/Project.tsx:37]: @jankeromnes, do you know the reason, why this should be triggered on both events? The question is also if there is a chance that this is run twice.

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (gtsiolis) [reply components/server/ee/src/gitlab/gitlab-app-support.ts:19]: Thanks @AlexTugarev! Seems 2x faster! FWIW, Still taking 2-3 seconds.

Comment (gtsiolis) [inline components/server/ee/src/gitlab/gitlab-app-support.ts:19]: **issue**: Tried on purpose logging in with GitHub only so that no there’s no active integration with GitLab and got the following error in console:

```js
Error: Request getProviderRepositoriesForUser failed with message: User xyz has no identity for host: gitlab.com!

File: components/server/ee/src/container-module.ts
Comment (AlexTugarev) [reply components/server/ee/src/container-module.ts:73]: `Branches` should be. 
`Prebuilds` are not expected in this change-set.

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (AlexTugarev) [reply components/server/ee/src/gitlab/gitlab-app-support.ts:19]: Awesome. How many projects do you see there? According to the logs, each API response takes ~600ms, which is really slow.

File: components/dashboard/src/projects/NewProject.tsx
Comment (AlexTugarev) [reply components/dashboard/src/projects/NewProject.tsx:277]: Let's go with the plain link below. 👍🏻 

Redesign is somewhat in conflict with the update to the design in general.

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (AlexTugarev) [reply components/server/ee/src/gitlab/gitlab-app-support.ts:19]: Nice! Thanks! `UserContext` isn't updated properly.

File: components/dashboard/src/projects/NewProject.tsx
Comment (gtsiolis) [reply components/dashboard/src/projects/NewProject.tsx:277]: DEAL 🤝

File: components/server/ee/src/container-module.ts
Comment (gtsiolis) [reply components/server/ee/src/container-module.ts:73]: OK! For some projects branches load but showing `UNKNOWN COMMIT MESSAGE` while others fail to load branches with the following error in the console:

```js
Uncaught (in promise) Error: Request getProjectOverview failed with message: Cannot read property 'avatar_url' of undefined
```

| Projecr Branches List |
|-|
| Screenshot 2021-08-17 at 10 24 54 AM |

File: components/server/ee/src/gitlab/gitlab-app-support.ts
Comment (gtsiolis) [reply components/server/ee/src/gitlab/gitlab-app-support.ts:19]: Most groups contain less than 10 projects. For the personal account _all_ projects (38) are listed.

**thought**: I think the delay happens while fetching all groups or the ones with proper permissions and maybe because the account is a direct member of the `gitlab-org` group which contains 300+ second-level groups. Tried this with a test account and groups and projects loaded in under 1 second. 🎉

File: components/server/ee/src/container-module.ts
Comment (AlexTugarev) [reply components/server/ee/src/container-module.ts:73]: `UNKNOWN COMMIT MESSAGE` that's a placeholder.

Comment (gtsiolis) [reply components/server/ee/src/container-module.ts:73]: Oh, ok! 🙈

File: components/dashboard/src/projects/NewProject.tsx
Comment (AlexTugarev) [reply components/dashboard/src/projects/NewProject.tsx]: "Unfortunately" this didn't work out, so I "had" to rewrite some parts of the the initial page in order to show the "No Access" within the component (the one with the dropdown.) Otherwise it's hard to decide what to render when "gitlab.com" is selected. Also if you are connected with gitlab.com, this will be the new default. 

Also, there is a new state: "loading..." ;-)

**Comment by gtsiolis** [comment]
Looking at this now! 👀

File: components/dashboard/src/projects/NewProject.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/NewProject.tsx:304]: **nitpick**: Minor color shift and small hack for improving the user experience while loading time, until we have a better loading indicator like the spinner described in https://github.com/gitpod-io/gitpod/issues/4948#issuecomment-895927769.

```suggestion
                        <h3 className="mb-2 text-gray-400 dark:text-gray-600 animate-pulse">
```

File: components/dashboard/src/projects/NewProject.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/NewProject.tsx:374]: **praise**: Nice touch, until we get the dropdown in place! 🎖️

Comment (gtsiolis) [inline components/dashboard/src/projects/NewProject.tsx:374]: **question**: Worth linking to the provider itself? Feel free to ignore if it does not seem helpful.

File: components/dashboard/src/projects/NewProject.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/NewProject.tsx:233]: **issue**: Is it expected that group avatars fail to load for groups? It seems that shared projects with write access on personal accounts successfully fetch the avatars.

| Missing avatars on groups (a) | Missing avatars on personal accounts (b)|
|-|-|
| Screenshot 2021-08-17 at 5 44 31 PM | Frame 34 |

File: components/dashboard/src/projects/NewProject.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/NewProject.tsx:227]: **question:** For some reason, the pre-selected account is a specific but maybe random group. Does it make sense to default to pre-selecting your own account?

File: components/dashboard/src/projects/NewProject.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/NewProject.tsx:305]: **issue**: The flow now stops with a _Loading ..._ placeholder when there's no active GitLab connection. Expected? There's also _**no**_ error in console. Should we add a flow step to authorize ad-hoc with GitLab.com as we do with users opening workspaces for repository from non-authorized providers, see https://github.com/gitpod-io/gitpod/issues/4197? Sounds good also as a follow up issue if needed. 🏓

**Review by gtsiolis — COMMENTED** [review]
Thanks @AlexTugarev! 🦊 

Did another pass and left some comments below. 🏀

File: components/dashboard/src/projects/NewProject.tsx
Comment (AlexTugarev) [reply components/dashboard/src/projects/NewProject.tsx:233]: indeed :-(

File: components/dashboard/src/projects/NewProject.tsx
Comment (AlexTugarev) [reply components/dashboard/src/projects/NewProject.tsx:227]: defaulting to own account sounds good.

File: components/dashboard/src/projects/NewProject.tsx
Comment (AlexTugarev) [reply components/dashboard/src/projects/NewProject.tsx:305]: reproduce and describe how that happened, please? it's a combination of effect, I suspect. 

the logic is: if gitlab.com is connected, use that, otherwise if github.com is connected, use that, otherwise nothing (unexpected, right?) we can remove this dead end by showing the git provider selection. TODO

Comment (gtsiolis) [reply components/dashboard/src/projects/NewProject.tsx:305]: Agree on the logic here! 🤓 

Steps to reproduce:
1. Login using GitHub at [`/login`](https://at-gitlab-projects.staging.gitpod-dev.com/login).
2. Make sure there's no active integration with GitLab at [`/integrations`](https://at-gitlab-projects.staging.gitpod-dev.com/integrations).
3. Go to Projects at [`/projects`](https://at-gitlab-projects.staging.gitpod-dev.com/projects).
4. Click New Project button on the Projects empty state page.
5. Flow stops with the _Loading ..._ placeholder trying to select a git repository on github.com.

**question**: Oh, maybe this is because the GitHub app is not configured for this preview environment, right?

File: components/dashboard/src/projects/NewProject.tsx
Comment (gtsiolis) [reply components/dashboard/src/projects/NewProject.tsx:227]: DEAL. Let's do this here or open a follow up issue to fine tune this later! Your call. 📞

File: components/dashboard/src/projects/NewProject.tsx
Comment (AlexTugarev) [reply components/dashboard/src/projects/NewProject.tsx:305]: I could reproduce it. Indeed this happens to happen if the GH App is not configured, but it would also be the case when no app installations are found, thus a very relevant case. Next deployment should fix this.

Comment (gtsiolis) [reply components/dashboard/src/projects/NewProject.tsx:305]: Thanks, @AlexTugarev! 🌟 

Works as expected. FYI, switching between providers also includes the delay mentioned in https://github.com/gitpod-io/gitpod/pull/5120#discussion_r689670664 when switching back to GitLab. 🙈

File: components/dashboard/src/projects/NewProject.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/NewProject.tsx:237]: **question**: Could we also hide this search element when the user has not authorized with the provider yet?

Comment (AlexTugarev) [reply components/dashboard/src/projects/NewProject.tsx:237]: good idea

File: components/dashboard/src/projects/Project.tsx
Comment (jankeromnes) [reply components/dashboard/src/projects/Project.tsx:37]: > @jankeromnes, do you know the reason, why this should be triggered on both events?

- If `team` changes, `await getGitpodService().server.getTeamProjects(team.id)` below will return something else
- If `teams` change, maybe a team got deleted or so, so we should probably update again?

> The question is also if there is a chance that this is run twice.

I'm not sure, why?

Comment (AlexTugarev) [reply components/dashboard/src/projects/Project.tsx:37]: I think it's just sufficient to depend on `teams`, because `team` is derived directly from `location` and `teams` state. Also, `team` isn't a React state, is it? It might be just ignored. 🤷🏻‍♂️  Having said that, I've no longer any concerns about double execution. ✔️

Comment (jankeromnes) [reply components/dashboard/src/projects/Project.tsx:37]: Is it possible that `teams` doesn't change, but the currently selected `team` changes? (E.g. based on the URL, or if the user opens a non-team project?)

**Comment by gtsiolis** [comment]
Looking at this now! 👀

**Comment by AlexTugarev** [comment]
### Known issues
- [ ] sticky repositories in list after account/group selection
   Screen Shot 2021-08-19 at 14 23 23
- [ ] `createProject failed with message: Cannot read property 'access_level' of null` when selecting a repo within a group

File: components/dashboard/src/projects/NewProject.tsx
Comment (gtsiolis) [inline components/dashboard/src/projects/NewProject.tsx:247]: **suggestion**: What do you think of using a smaller max height like `max-h-52` here? Suggesting this mostly because the scolling area is overflowing below the viewport height in some cases. Your call! We can also pick up such details in https://github.com/gitpod-io/gitpod/issues/4948 and https://github.com/gitpod-io/gitpod/issues/4951.

| BEFORE | AFTER |
|-|-|
| Screenshot 2021-08-19 at 5 31 45 PM | Screenshot 2021-08-19 at 5 31 23 PM |

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **alextugarev** after the PR has been reviewed.

*No associated issue*. Update pull-request body to add a reference to an issue, or get approval with `/approve no-issue`

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)**
- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Review by gtsiolis — COMMENTED** [review]
Thanks @AlexTugarev! Left a comment about the repository selection step.

**question**: Did we remove the loading placeholder?

**Comment by gtsiolis** [comment]
@alexleventer Also, opened an issue about the missing default branch label and sorting the default branch on top in https://github.com/gitpod-io/gitpod/issues/5278, however, feel free to leave that out of the scope of these changes.

**Comment by AlexTugarev** [comment]
/hold

... to get a change to squash the commits.

**Comment by AlexTugarev** [comment]
> question: Did we remove the loading placeholder?

@gtsiolis, thanks for the reminder! first I thought it got so fast, that it's no longer needed ;-) 
It should be back in with the last update.

**Comment by AlexTugarev** [comment]
Known issue: wrong width for the bottom section. 

Screen Shot 2021-08-19 at 17 52 32

@gtsiolis, wdyt of getting rid of the grayish background and render as simple as in https://github.com/gitpod-io/gitpod/pull/5120#discussion_r685405180?

**Comment by jankeromnes** [comment]
Many thanks for adding GitLab support! 💯 🎆 

I briefly skimmed the code, but I'm unfamiliar with the GitLab integration code and didn't have time to read more deeply into it.

I also briefly tested this PR:

1. Connected both GitHub and GitLab accounts
2. Tried to create a new project for my user
3. There is a new loading state before anything shows up (empty list felt better)
4. It shows me this -- it feels I don't get the choice between GitHub or GitLab?

Screenshot 2021-08-19 at 18 10 19

5. Adding one of the proposed GitLab project seems to work well 🎉 (the old Configurator is even able to fetch the config)

Screenshot 2021-08-19 at 18 14 50

6. Trying again a few times, I randomly stumbled on this hidden option to add GitHub projects and not GitLab -- feels like this should be placed somewhere much more obvious

Screenshot 2021-08-19 at 18 16 10

7. Authorizing GitHub didn't work (but I guess that's expected since it's not the focus of this PR -- still would have been cool to test GitLab and GitHub projects together)

Screenshot 2021-08-19 at 18 17 49

8. It doesn't even take that long to import the GitLab repo itself 💪 (took barely a few seconds to load the config)

Screenshot 2021-08-19 at 18 19 06

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **alextugarev** after the PR has been reviewed.

Associated issue: *#5278*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)**
- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/server/src/projects/projects-service.ts
Comment (svenefftinge) [inline components/server/src/projects/projects-service.ts:104]: It should work with self-managed gitlab's as well.

**Review by svenefftinge — APPROVED** [review]
Looks good and works well with gitlab.com. From the code besides one place it looks like it should work with self-managed GitLab's too.

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: 22ca396613977b4cf0d10117e4d89c57daa1e41a</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/5120#pullrequestreview-734987445" title="Approved">svenefftinge</a>*

Associated issue: *#5278*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)~~ [svenefftinge]
- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [svenefftinge]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [svenefftinge]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by AlexTugarev** [comment]
> From the code besides one place it looks like it should work with self-managed GitLab's too.

@svenefftinge, have you tested prebuilds with self-managed GitLabs?

File: components/server/src/projects/projects-service.ts
Comment (gtsiolis) [reply components/server/src/projects/projects-service.ts:104]: **fyi**: There's a follow up issue for self hosted support in https://github.com/gitpod-io/gitpod/issues/5115. Cc @svenefftinge

File: components/dashboard/src/projects/NewProject.tsx
Comment (gtsiolis) [reply components/dashboard/src/projects/NewProject.tsx:222]: **fyi**: Opened follow up issue in https://github.com/gitpod-io/gitpod/issues/5330.

File: components/dashboard/src/projects/NewProject.tsx
Comment (gtsiolis) [reply components/dashboard/src/projects/NewProject.tsx:228]: **fyi**: Opened follow up issue in https://github.com/gitpod-io/gitpod/issues/5330.

File: components/dashboard/src/projects/NewProject.tsx
Comment (gtsiolis) [reply components/dashboard/src/projects/NewProject.tsx:222]: **fyi**: Opened follow up issue in https://github.com/gitpod-io/gitpod/issues/5331.

File: components/dashboard/src/projects/NewProject.tsx
Comment (gtsiolis) [reply components/dashboard/src/projects/NewProject.tsx:371]: **fyi**: Opened follow up issue in https://github.com/gitpod-io/gitpod/issues/5333

File: components/dashboard/src/projects/NewProject.tsx
Comment (gtsiolis) [reply components/dashboard/src/projects/NewProject.tsx:242]: **fyi**: Opened follow up issue in https://github.com/gitpod-io/gitpod/issues/5334.

**Pull request merged**
